### ISSUE TITLE
Critical Bug Fix: CV Output is not tracking correctly.

### ISSUE THREAD

**Issue by jotel-california** [Befaco/MIDIThing2#43]
Using the newest 1.2 Test 3 Firmware.

I always wondered why my sequences sometimes sound out of tune, and here is my answer. The midi thing does not track correctly. 
If I tune Midi Note zero to a C, I get substantial errors. 
1st oct = -15ct
2nd oct = -20ct
3rd oct -26
4rd oct -28
5th oct -29
6th octave -32ct

I tried 3 other CV sources, and all other sources track correctly, so It's definitly the midi thing
I hope this can be fixed in software.

**Comment by jotel-california** [comment]
here is a demonstration with video: https://youtu.be/M3KUzgGkbFM

**Comment by jjobse** [comment]
Hello. Have you tried measuring the note voltage straight from the midi thing v2? I had a similar issue but it was my oscillator not tracking properly. FWIW I have seen very stable voltage across multiple octaves from the midi thing v2. Most oscillators have a calibration trim pot on the back you can tweak to resolve this issue. Also once you confirm there is a 1v different in voltage across each octave make sure you let the oscillator warm up for 15mins or more before you start measuring. I used my Zoia to measure the voltage and the pitch ranges with great success.

**Comment by jotel-california** [comment]
Hey. Yes I did, and just plugging the midi thing into the multimeter and there the voltages read ok. Although „0V“ actually is 0.2V on the midi thing. Then it scales correctly, 1 oct above is 1.2 and so on. However as soon as you plug it into anything it stops tracking correctly. I can rule out any oscillator flaws 100%.I tested it with 6 different oscillators across 2 systems with different PSUs, analog as well as digital oscs. The behaviour is the same, so it is 100% the midi thing. Are the outputs buffered? Cause if not, the impedance if the cv in could distort tracking. JoAm 20.11.2024 um 14:53 schrieb jjobse ***@***.***>:﻿
Hello. Have you tried measuring the note voltage straight from the midi thing v2? I had a similar issue but it was my oscillator not tracking properly. FWIW I have seen very stable voltage across multiple octaves from the midi thing v2. Most oscillators have a calibration trim pot on the back you can tweak to resolve this issue. Also once you confirm there is a 1v different in voltage across each octave make sure you let the oscillator warm up for 15mins or more before you start measuring. I used my Zoia to measure the voltage and the pitch ranges with great success.

**Comment by jjobse** [comment]
Ah interesting ok I’m tracking. Yea mine is not exactly 0vs as well but does track 1v per octave. Do you see the same behavior if you stick a buffered mult in between the midi thing and the oscillator? One of my older Pittsburgh modular midi and oscillator combos had that same issue and I had to send the midi device in for repair/replacement. 
Sorry to hear about the issues hopefully you can get it sorted out. This is my first befaco product and I am a little concerned with all the bugs I see with it (especially with such a high price) so I have been checking this page constantly for updates. I love the screen, the way you can configure it to work with many different oscillator types, and the number of outputs (I can get poly into the rack and multiple mono voices at the same time). But this is one of the most bug filled eurorack devices I have ever dealt with and it’s missing some pretty basic features. Makes me wish my zoia had an expansion bay.

**Comment by jotel-california** [comment]
Yes, a buffered mult does not help. Sure, the extreme voltage drop when passively splitting the cv is gone, but it still does not track exactly.Am 20.11.2024 um 18:03 schrieb jjobse ***@***.***>:﻿
Ah interesting ok I’m tracking. Yea mine is not exactly 0vs as well but does track 1v per octave. Do you see the same behavior if you stick a buffered mult in between the midi thing and the oscillator? One of my older Pittsburgh modular midi and oscillator combos had that same issue and I had to send the midi device in for repair/replacement.
Sorry to hear about the issues hopefully you can get it sorted out. This is my first befaco product and I am a little concerned with all the bugs I see with it (especially with such a high price) so I have been checking this page constantly for updates. I love the screen, the way you can configure it to work with many different oscillator types, and the number of outputs (I can get poly into the rack and multiple mono voices at the same time). But this is one of the most bug filled eurorack devices I have ever dealt with and it’s missing some pretty basic features. Makes me wish my zoia had an expansion bay.

**Comment by xustafu** [comment]
> Using the newest 1.2 Test 3 Firmware.
> 
> I always wondered why my sequences sometimes sound out of tune, and here is my answer. The midi thing does not track correctly. If I tune Midi Note zero to a C, I get substantial errors. 1st oct = -15ct 2nd oct = -20ct 3rd oct -26 4rd oct -28 5th oct -29 6th octave -32ct
> 
> I tried 3 other CV sources, and all other sources track correctly, so It's definitly the midi thing I hope this can be fixed in software.

Tested on a couple of units and could see a total deviation of 10 cents. Which is not perfect but pretty decent.
Did you tried with other VCOs? is this 29 cents deviation constant?

Anyway, allow us a couple of  days to cook a calibration procedure so you can test on your side. We will come back with a test firmware for you!

**Comment by xustafu** [comment]
>  But this is one of the most bug filled eurorack devices I have ever dealt with and it’s missing some pretty basic features. 

Sorry for such a dispointing user experiece.
Please feel free to keep on reporting bugs and we will do our best to fix them all. Thansk a lot for your valuable feedback.

**Comment by jotel-california** [comment]
> > Using the newest 1.2 Test 3 Firmware.
> > I always wondered why my sequences sometimes sound out of tune, and here is my answer. The midi thing does not track correctly. If I tune Midi Note zero to a C, I get substantial errors. 1st oct = -15ct 2nd oct = -20ct 3rd oct -26 4rd oct -28 5th oct -29 6th octave -32ct
> > I tried 3 other CV sources, and all other sources track correctly, so It's definitly the midi thing I hope this can be fixed in software.
> 
> Tested on a couple of units and could see a total deviation of 10 cents. Which is not perfect but pretty decent. Did you tried with other VCOs? is this 29 cents deviation constant?
> 
> Anyway, allow us a couple of days to cook a calibration procedure so you can test on your side. We will come back with a test firmware for you!

Alright, excited to see what you can come up with!

the tracking error is a little bit different with different vcos. Tried it with 6 different oscs in 2 different systems with different psus. 
digital aswell as analog. 
But it definitly is similar so maybe not -29ct, but -22ct or similar but I guess that could really be the tracking error of the oscillators themselves. 
Or one other thought I had is that if the outputs of the midi thing are unbuffered, the impedance of the cv inputs could affect the output level.

**Comment by xustafu** [comment]
@jotel-california  Here you go!
[MidiThing2_calibra.zip](https://github.com/user-attachments/files/17921211/MidiThing2_calibra.zip)

Calibration pocedure is available in the Options menu of the CV option of a voice:

![image](https://github.com/user-attachments/assets/7dba9741-f84e-47e9-8444-65940d9baf2f)

We will be calibrating two points, so the slope of the conversion will be adjusted. This will correct tuning deviations with connected VCO for the selected port. 
This calibration affects only the CV port selected. Keep in mind this calibration is not global. 
Be warned, changing voltage or notes Range will override the calibration, so change those parameters before calibrating.

The procedure is:

Connect your oscillator to a Tuner, making sure it is tuned exactly at a certain note (any note in 0 octave should be good).
Connect the CV output in MIDI thing 2 to your Oscillator V/oct. IN. 
Navigate to CV Options menu:

- Send MIDI Note 24 (that should correspond to 2v) with your preferred MIDI controller. Go to 2v cent adjustment A. Tweak it until your oscillator is exactly two octaves over the initial point. The numbers will be the amount of cents you are moving the CV up or Down the original value.
- Send MIDI Note 96 now  (that should correspond to 8v) Go to 8v cent adjustment B. Tweak it until your oscillator is exactly eight octaves over the initial point. The numbers will be the amount of cents you are moving the CV up or Down the original value.
![image](https://github.com/user-attachments/assets/84ba48d8-7357-4b6d-af13-6ab8a71f45f6)

Now this CV out should have improved tracking linearity! Save the configuration in order to keep the changes!

WARNING: these steps might change in future versions. Please make sure to read docuemtnation for your own version

**Comment by jotel-california** [comment]
Wow that was quick. Amazing work. Just tested it, it works! I can now make my vcos track with about 6ct precision,  That's more than enough and makes scaling different oscillators a breeze. 
That's a huge helper. You also satisfied the other guy who wanted 1.2v scaling lol.

One thing: I would recommend, is making the scale smaller. 1 adjustment is often too much, if you want to tune it really precisely. I think 4 times the resolution would be fine.

Spotted a bug: If the scaling is changed, playing the lowest note (0V Point, so Midi Note 0 in the default setting) makes the midi thing output the maximum voltage possible, in this case 10V instead of 0V.  
However, if you change the voltage range, lets say from 0-10V to 0-8V, this behaviour stops. Midi Note 0 is 0V again. However, now the scaling setting is ignored (even though the value on the screen is maintained!!) until I change the scaling parameter again. Then however, it is back to Midi note 0 = 10V behaviour. 

I get that for different voltage ranges the calibration procedure needs to be re-done, but I'd prefer to keep the calibration numbers when changing voltage range, even though the outcome might be incorrect.
Other possibility would be to reset the calibration to 0 when you change voltage range. 

If the scaling is off, 0V still has the greatest error. it's always around -12ct compared to 1V. 
1-9V on the oher hand works better.

**Comment by jotel-california** [comment]
I meant +12ct on the 0V point, sorry.
